### ISSUE TITLE
Refactor the Plot Module

### ISSUE THREAD

**Issue by StendProg** [swharden/ScottPlot#573]
**What can we do to make ScottPlot better?**

Current `Plot.cs` source file contain about 2500 lines of code. Each time working with such a huge file is difficult.
Grouping by regions only partially solves the problem. Most of the code is API from which there is no escape.

It may make sense to split the current class into different files, at least we can move the `PlotScatter, PlotSignal, etc` API definitions to a separate file, or even better to a group of files.

Plot.cs
```c#
public partial class Plot
{
 // all definitions without Plottables API
  ....
}
```
PlotPlottablesAPI.cs
```c#
public partial class Plot
{
  public PlottableScatter PlotScatter(...)
  {
   ...
  }
  public PlottableSignal PlotSignal(...)
  {
    ...
  }
 ...
}

**Comment by swharden** [comment]
Hi @StendProg, thank you for this suggestion! This topic is one I've been thinking a lot about over the last few months.

### The Problem

I think everyone agrees that the `Plot` module has grown far too large. Most of its size comes from methods which simply create plottables (using optional arguments) and add them to the list. The core function of the `Plot` module is diluted by these helper functions. 

### Partial Classes

Above you suggest splitting the `Plot` module into smaller files using partial classes. I like this idea - it would reduce file size and help isolate concerns by putting all current plottable creation helper methods in one file. However, in researching this topic I found that using partial classes is typically not advised because if a class requires partial classes it means that class is too large. 

I do think we should make the change you propose. However, I also think the `Plot` module remains too large, and additional options should be explored...

### New Method to Create Plots?

Something I've been strongly considering, as soon as ScottPlot 4.1, is to mark all of the current plot creation methods obsolete. They would still work, but users would be advised to no longer use them. This would be easier if they were all isolated in a partial class file. 

Simultaneously we would remove all styling arguments from plottable constructors, and plottables would only accept data variables as constructor arguments. Like the `Plot` module, plottable constructors have grown far too large. Customizing styling or behavior would be achieved by assigning to public properties of the plottables after they are created. This change would be associated with some changes in namespaces to make plottables more easily discoverable. The way users would create plots would be:

* Create a Plot
* Create a Plottable (passing-in data)
* Optionally customize the Plottable (using public properties)
* Add the Plottable to the Plot

Code would look something like:
```cs
var plt = new Plot();
var scatter = new Plottable.Scatter(xs, ys) { Color = Color.Red, MarkerSize = 3 };
plt.Add(scatter);
```

There are a lot of small considerations, edge cases, and implications this new sequence would have. It definitely would reduce coupling and improve modularity, but it would certainly be a big change for people used to the style promoted by ScottPlot 4.0. I have been considering exploring this in ScottPlot 4.1, and marking it as alpha so it wouldn't surprise anyone upgrading with NuGet.

@StendProg I'm interested in your thoughts here! This seems like a big response to your initial suggestion, but 4.0 -> 4.1 may be a good time to consider it...

**Comment by StendProg** [comment]
Hi @swharden,
> is typically not advised because if a class requires partial classes it means that class is too large

I think having a lot of API related methods allows this rule to be broken.

>New Method to Create Plots?

```c#
var scatter = new Plottable.Scatter(xs, ys) { Color = Color.Red, MarkerSize = 3 };
plt.Add(scatter);
```
This is already working now (not through initializers but through constructor parameters), but compare to 
```c#
plot.PlotScatter(xs, ys, Color: Color.Red, MarkerSize: 3);
```
The new method requires exactly 2 times as many lines of code, I think this is a significant drawback.

Nobody bothers to change the fields for properties and simplify constructors, I am all for it. But in parallel it is necessary to leave the existing API (`plot.PlotPlottable() methods`), it is very successful, especially since it has already been written. It is really very attractive to new users (especially those familiar with python as far as I can imagine). It is wrong to solve developer problems at the expense of the user.


```c#
var scatter = plot.Plot(new Plottable.Scatter(xs, ys){ Color = Color.Red});
```
It looks attractive, but still worse.

One way or another, no one bothers to split the class right now, and somehow then come up with something better.

**Comment by swharden** [comment]
Hi @StendProg, thank you for your insightful feedback! It has been very helpful discussing these topics with you.

### Partial Classes

> > is typically not advised because if a class requires partial classes it means that class is too large

> I think having a lot of API related methods allows this rule to be broken.

I like this conclusion. Typically classes should be small, but since this is the primary user-facing API it can be large. The files appear especially large because every argument is given its own line. **I will split the Plot module like you suggest.** I will probably name this file `PlotAddAPI.cs`.

However, there is an alternative way to split this class... What if we put partial classes inside the files that define each plottable? The top of `PlottableScatter.cs` would have a small partial class to extend the `Plot` module with a method to create a scatter plot. This means adding new customizations to a plottable could be achieved by editing a single file, rather than editing the plottable file and also a large API file in another folder. I could begin implementing this change, one plottable at a time, with no breaking changes. **What do you think of this strategy?**

### Plot Module Structure

> The new method requires exactly 2 times as many lines of code, I think this is a significant drawback.

> it is necessary to leave the existing API ... since it has already been written. 

> It is really very attractive to new users (especially those familiar with python as far as I can imagine). 

> It is wrong to solve developer problems at the expense of the user

These are all excellent points, and your message is very well-received. **I will leave the existing methods are they are.**

### Simplifying Plottable Constructors

I still like the idea of refactoring to simplify the constructors of plottables so that they accept only data, and public properties customize styling and behavior. Changes like this will involve changing code inside methods in the plot module, and are unlikely to affect users (since users typically do not construct plottables themselves). This improvement will reduce complexity of plottables and have no effect on users, so it seems like it has benefits and no disadvantages.

**Comment by StendProg** [comment]
Hi @swharden,

Dropped out for a couple of days, sorry.
It looks like you yourself decided everything in the best way.
> Plot.AddPlottable.cs

It couldn't be better!

The only unresolved question is, if you dump all `PlotPlotable` into one file, it will still turn out to be huge, it would be great to split all existing Plotables into 3-5 logically related groups and implement the api of each of these groups in a separate file.

If you could help me with the choice of group names, I could do a PR with everything else. Or do you have some other plans in this regard? The current state looks like it is stuck halfway.

If we think about the ideal solution of this Issue. If each `PlotPlottable`api method did nothing but create a Plottable and adjust its parameters, then we could try to generate the code of these API methods automatically according to some simple rules. Unfortunately, now some AddPlotables have very different functionality, and I also vaguely imagine how to implement code generation.

**Comment by swharden** [comment]
> Dropped out for a couple of days, sorry.

No worries!

> It looks like you yourself decided everything in the best way.

I'm glad you agree with this direction 👍 

> if you dump all PlotPlotable into one file, it will still turn out to be huge, it would be great to split all existing Plotables into 3-5 logically related groups and implement the api of each of these groups in a separate file.

I like this idea! Perhaps we can break it into these groups. If we do this, perhaps we should move `Plot.cs` into `Plot/Plot.cs` and put all these files in that folder (without changing the namespace) to keep them together.
* `Plot.PlotBar.cs` - population, bar
* `Plot.PlotFinance.cs` - candle, ohlc
* `Plot.PlotScatter.cs` - scatter, point, line
* `Plot.PlotSignal.cs` - Signal, SignalXY, SignalConst
* `Plot.PlotBitmap.cs` - image, heatmap
* `Plot.PlotText.cs` - text, annotation
* `Plot.PlotAxisLine` - hline, vline, hspan, vspan
* `Plot.PlotMisc.cs` - errorbar, polygon, function, pie, radar, vector field

> I could do a PR with everything else. Or do you have some other plans in this regard? The current state looks like it is stuck halfway.

Thank you for your offer, but I think it would be a good idea for me to do this. It has been a long time since I have looked at many of these methods and I would prefer to this carefully and slowly, refining each method one-by-one. I also want to make this change as I refactor the constructor of plottables to remove styling arguments.

> The current state looks like it is stuck halfway.

Indeed! I'm experimenting with the best way to refactor plottable constructors, and I am only working on the bar plot until I figure out the best way. I hope to get the bar plot looking good in the next few days, then refactor the rest of the plottables to use a similar style.

> we could try to generate the code of these API methods automatically 

This may be possible, but I prefer the simplicity of having a few hand-crafted methods. If there were >100 plot types I would feel differently, but there are so few that I think it is easy to maintain them by hand.

I will continue to work on this over the weekend! Thank you @StendProg, as always, for your valuable input.

**Comment by swharden** [comment]
I broke the plot module up into more pieces than I initially anticipated. In addition to partial class files for plots, it seemed useful to create additional files for ticks, styling, settings (behavior), axis (coordinate lookups), and axis labels.

Now the Plot module is less than 250 lines! However, it is now much more obvious that the rendering system is very ugly. This is probably a good thing, because it makes it much more obvious that I need to continue to work to refactor the rendering system. Having all those extra styling, configuration, and plottable methods outside the Plot.cs file makes this task less intimidating.

Originally I thought I would refactor constructors of all plottables as I did this work, but I decided to go ahead and do this work now and treat the constructors as a separate issue. That refactor will be easier since all those methods are now in smaller files.

Thank you for your original suggestion @StendProg!
